src/ipred.rs

```diff
@@ -607,33 +607,28 @@ unsafe extern "C" fn ipred_paeth_c_erased<BD: BitDepth>(
 unsafe fn ipred_smooth_rust<BD: BitDepth>(
     mut dst: *mut BD::Pixel,
     stride: ptrdiff_t,
-    topleft: *const BD::Pixel,
+    topleft: &[BD::Pixel; SCRATCH_EDGE_LEN],
+    topleft_off: usize,
     width: c_int,
     height: c_int,
-    _a: c_int,
-    _max_width: c_int,
-    _max_height: c_int,
-    _bd: BD,
 ) {
-    let weights_hor: *const u8 = &*dav1d_sm_weights.0.as_ptr().offset(width as isize) as *const u8;
-    let weights_ver: *const u8 = &*dav1d_sm_weights.0.as_ptr().offset(height as isize) as *const u8;
-    let right = (*topleft.offset(width as isize)).as_::<c_int>();
-    let bottom = (*topleft.offset(-height as isize)).as_::<c_int>();
-    let mut y = 0;
-    while y < height {
-        let mut x = 0;
-        while x < width {
-            let pred = *weights_ver.offset(y as isize) as c_int
-                * (*topleft.offset((1 + x) as isize)).as_::<c_int>()
-                + (256 - *weights_ver.offset(y as isize) as c_int) * bottom
-                + *weights_hor.offset(x as isize) as c_int
-                    * (*topleft.offset(-(1 + y) as isize)).as_::<c_int>()
-                + (256 - *weights_hor.offset(x as isize) as c_int) * right;
-            *dst.offset(x as isize) = (pred + 256 >> 9).as_::<BD::Pixel>();
-            x += 1;
+    let [width, height] = [width, height].map(|it| it as usize);
+
+    let weights_hor = &dav1d_sm_weights.0[width..];
+    let weights_ver = &dav1d_sm_weights.0[height..];
```

Comment:
```suggestion
    let weights_hor = &dav1d_sm_weights.0[width..][..width];
    let weights_ver = &dav1d_sm_weights.0[height..][..height];
```

To only do the bounds checks once.